### PR TITLE
ENG-560: Atomic Bridge Shared Crate

### PR DESCRIPTION
**Pull request by cde8**
# Summary
- **RFCs**: RFC040 Atomic bridge
- **Categories**: protocol-units

## Structure
Implemented the logic for managing "structure" by utilizing the atomic swap primitive. This shared crate incorporates a rigorously tested architecture for transferring assets between two blockchains through trait interfaces.

It is presumed that asset transfer is feasible when there are contracts present on both sides of the blockchain:

Blockchain 1:
Initiator
Counterparty

Blockchain 2:
Initiator
Counterparty

Blockchain 1 Initiator will work with Blockchain 2 Counterparty

and

Blockchain 2 Initiator will work with Blockchain 2 Counterparty

## Conversions
In scenarios involving multiple blockchains and distinct Rust blockchain clients, variations in hash and address formats are encountered. To address this, developers using this crate are required to encapsulate hash and address types and define traits for facilitating conversion, ensuring equality comparisons through PartialEq and Eq implementations. This approach enables seamless comparison of addresses and hashes across diverse blockchain environments.

## Testing
Thorough emphasis is placed on rigorously testing the bridging functionality, recognizing its pivotal role in ensuring operational integrity. I have developed an AbstractBlockchain framework that enables manipulation of time, facilitating comprehensive testing scenarios. These tests are designed to simulate diverse client behaviors, including false positives and random errors, enhancing reproducibility for debugging purposes through the use of a seed. This approach enables efficient detection of abnormalities within the bridging functionality, effectively simulating prolonged operational conditions in a condensed timeframe.

## Multiple sources of trust
Diversifying sources of trust is a critical element in the context of bridging, as the bridging service will interact with events that are consumed by client libraries. By ensuring the implementation of the contract and monitoring traits, the integration of multiple sources of truth can effectively function in conjunction with the respective client libraries.

## Integration
Initially, the integration of Ethereum and Aptos-like chains will involve the implementation of two additional crates utilizing their respective client libraries. These clients must adhere to the corresponding traits, namely BridgeInitiatorContract and BridgeCounterpartyContract. Furthermore, it is essential to implement the monitoring traits. Upon successful implementation of these traits within the crates, the BridgingService structure can be properly configured, resulting in a functional bridging service. This setup will also facilitate seamless expansion to other blockchain networks.

# Changelog
1. Added bridge-cli, bridge-service, and bridge-shared packages to handle different aspects of blockchain bridging.
2. Updated Cargo.toml files to include new workspace members for bridge protocol units.
3. Introduced new dependencies in Cargo.lock for added packages ensuring compatibility across asynchronous tasks and cryptographic functions.
4. Created main executables for CLI and service units with initial greeting message.
5. Developed foundational structures for managing blockchain events and interactions within the bridge-shared unit.

# Testing
- Ensured all unit tests cover the new functionalities introduced in the bridge-cli, bridge-service, and bridge-shared units.
- Added integration tests to simulate real-world transaction scenarios across blockchain services.
- Performed regression tests to confirm that existing functionality remains unaffected by the new updates.

# Outstanding issues
- Further integration testing with external blockchain services to guarantee robust performance under various network conditions.
- Expansion of test coverage to include error handling and failure scenarios.

### CODE REVIEW

File: protocol-units/bridge/shared/tests/bridge_service_error_handling.rs
Comment (andygolay) [inline protocol-units/bridge/shared/tests/bridge_service_error_handling.rs:1]: This all looks amazing, @cde8. One small comment / question: I notice you have this error handling file, and you also have `bridge_service/error_handling.rs`. Would it make sense to organize them together on the same level, as they both appear to deal with bridge service error handling?

File: protocol-units/bridge/cli/src/main.rs
Comment (l-monninger) [inline protocol-units/bridge/cli/src/main.rs:1]: Was this meant to be left in?

File: protocol-units/bridge/shared/src/blockchain_service.rs
Comment (l-monninger) [inline protocol-units/bridge/shared/src/blockchain_service.rs:62]: Can we get doc comments on these?

File: protocol-units/bridge/shared/src/bridge_service/active_swap.rs
Comment (l-monninger) [inline protocol-units/bridge/shared/src/bridge_service/active_swap.rs:201]: Can you explain the need in comments to implement our own wakers/runtime?

File: protocol-units/bridge/shared/tests/abstract_blockchain.rs
Comment (l-monninger) [inline protocol-units/bridge/shared/tests/abstract_blockchain.rs:74]: I've commented this on other folks PRs. But, I prefer that we avoid unwraps even in tests. In tests because people often use tests as examples and copy code accordingly.

File: protocol-units/bridge/shared/tests/blockchain_service.rs
Comment (l-monninger) [inline protocol-units/bridge/shared/tests/blockchain_service.rs:31]: Again, it seems like our own runtime might be overkill here. If we're doing this for some kind ordering, shouldn't the assumption be that the `BlockchainService` will (and in fact must) provide its own ordering.

File: protocol-units/bridge/shared/tests/bridge_service.rs
Comment (l-monninger) [inline protocol-units/bridge/shared/tests/bridge_service.rs:122]: "Preimage" is one word in mathematics and cryptography. I don't think the camel-case is needed here.

**Review by l-monninger — CHANGES_REQUESTED** [review]

**Comment by l-monninger** [comment]
There aren't any concrete client implementations yet, right? @cde8

File: protocol-units/bridge/cli/src/main.rs
Comment (cde8) [reply protocol-units/bridge/cli/src/main.rs:1]: Hey Liam,  this was intentionally included. Think of it as a temporary placeholder for the to be developed bridge command line interface.

**Comment by l-monninger** [comment]
Overall, there's a lot of abstraction here and not a lot doc comments. I think to give a proper review I'd like to see more doc comments, particularly as pertain to the usage of various portions of the API.

File: protocol-units/bridge/shared/tests/bridge_service_error_handling.rs
Comment (cde8) [reply protocol-units/bridge/shared/tests/bridge_service_error_handling.rs:1]: Great find @andygolay I included this as a rough sketch for the error handling process. I will remove it.

Comment (cde8) [reply protocol-units/bridge/shared/tests/bridge_service_error_handling.rs:1]: This is more a test of the client interface, as it should fail with an invalid secret; this is actually assumed and could be removed completely for clarity. I will remove this file.

File: protocol-units/bridge/shared/src/bridge_service/active_swap.rs
Comment (cde8) [reply protocol-units/bridge/shared/src/bridge_service/active_swap.rs:201]: Could you clarify what you mean by implementing our own Waker or runtime?

File: protocol-units/bridge/shared/tests/abstract_blockchain.rs
Comment (cde8) [reply protocol-units/bridge/shared/tests/abstract_blockchain.rs:74]: I understand your perspective. The code in question is designed primarily for testing purposes, and it is not intended for broader use. The potential failure mode identified would predominantly occur due to insufficient memory. Converting this to an `expect` could be considered, although the necessity of such a change might be minimal. However, if you deem the modification essential, I am prepared to implement this adjustment. Please advise on how you wish to proceed.

File: protocol-units/bridge/shared/tests/blockchain_service.rs
Comment (cde8) [reply protocol-units/bridge/shared/tests/blockchain_service.rs:31]: I'm unsure about your references to "ordering" and "runtime," Liam. Could you clarify what you mean? However, I'm not clear on how this relates to implementing a runtime. There is no runtime implementation here. Only the check that items get cleaned up as they should. Could you please explain further?

File: protocol-units/bridge/shared/tests/bridge_service.rs
Comment (cde8) [reply protocol-units/bridge/shared/tests/bridge_service.rs:122]: I have included the case for clarity. Would you prefer it to be renamed to HashLockPreimage? The same logic could also be applied to a HashLock for example. You would also consider that one word. Please give me your preferences and I will rename it towards whatever you like.

**Comment by cde8** [comment]
> There aren't any concrete client implementations yet, right? @cde8

The concrete implementation will be done as soon as the movement and Aptos crates are developed, reviewed, and merged according to the original planning.

File: protocol-units/bridge/shared/src/bridge_service/active_swap.rs
Comment (l-monninger) [reply protocol-units/bridge/shared/src/bridge_service/active_swap.rs:201]: You have implemented what looks something like coroutines--with wakers and polls--and with very similar semantics to how you would implement async runtimes in rust.

File: protocol-units/bridge/shared/tests/abstract_blockchain.rs
Comment (l-monninger) [reply protocol-units/bridge/shared/tests/abstract_blockchain.rs:74]: I will comment it again on a later PR or put it in an issue. I suppose we're going ahead with the merge here to unblock @0xmovses

File: protocol-units/bridge/shared/tests/blockchain_service.rs
Comment (l-monninger) [reply protocol-units/bridge/shared/tests/blockchain_service.rs:31]: You're implementing poll and waker logic at a lower level than simply relying on async methods. Why?

File: protocol-units/bridge/shared/tests/bridge_service.rs
Comment (l-monninger) [reply protocol-units/bridge/shared/tests/bridge_service.rs:122]: It's a nit.

**Review by l-monninger — APPROVED** [review]

**Comment by l-monninger** [comment]
@cde8 On second review, I see why you chose to implement `Future` and `Stream` and release manipulations of the `Waker` seem to be confined to tests.

File: protocol-units/bridge/shared/tests/blockchain_service.rs
Comment (cde8) [reply protocol-units/bridge/shared/tests/blockchain_service.rs:31]: If you are aiming for a higher degree of control beyond the use of async methods, particularly for detailed testing, it becomes necessary to adopt a more nuanced implementation approach. 

This subject might warrant a thorough, dedicated discussion on design tradeoffs beyond the scope of a pull request.

**Pull request merged**
